From 9669454d44f9dcafe2d5fb7f046671d4f702c809 Mon Sep 17 00:00:00 2001
From: Mateusz Poliwczak <mpoliwczak34@gmail.com>
Date: Thu, 6 Nov 2025 07:52:01 +0100
Subject: [PATCH 045/133] std.ArrayList: actaully memset to undefined in
 shrinkRetainingCapacity and clearRetainingCapacity
X-Developer-Signature: v=1; a=openpgp-sha256; l=1807; i=xtex@astrafall.org;
 h=from:subject; bh=bis9z+K4PN1vNEzqyRsrir6U95L1FCTtdnI/8AnF/LY=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv/LMJ8+7xR94Eif6aJZnw94LppctkWcz/rTXH+Xe
 7YiJXt2dkxkYRDjYrAUU2QpMmzwZtVJ5xddVi4LM4eVCWSItEgDAwMDAwsDX25iXqmRjpGeqbah
 nqGhjpGOEQMXpwBMtY0Fw3/Py486GOPfztDIFZd+7Tfxu3CcvkXuw4SjqZe17CwPydQwMny8qDG
 z81ExQ+mS9LI5H2MD7mwUtY+zKn+0h49z8TwfCW4A
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

See #25810

(cherry picked from commit b2895f356fa6203d0f3f1476c4e8fb9cf5b6d860)
---
 lib/std/array_list.zig | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/lib/std/array_list.zig b/lib/std/array_list.zig
index ab5985685c14..2ec941dd95ab 100644
--- a/lib/std/array_list.zig
+++ b/lib/std/array_list.zig
@@ -413,15 +413,15 @@ pub fn AlignedManaged(comptime T: type, comptime alignment: ?mem.Alignment) type
         /// Asserts that the new length is less than or equal to the previous length.
         pub fn shrinkRetainingCapacity(self: *Self, new_len: usize) void {
             assert(new_len <= self.items.len);
-            self.items.len = new_len;
             @memset(self.items[new_len..], undefined);
+            self.items.len = new_len;
         }
 
         /// Reduce length to 0.
         /// Invalidates all element pointers.
         pub fn clearRetainingCapacity(self: *Self) void {
-            self.items.len = 0;
             @memset(self.items, undefined);
+            self.items.len = 0;
         }
 
         /// Invalidates all element pointers.
@@ -1185,11 +1185,14 @@ pub fn Aligned(comptime T: type, comptime alignment: ?mem.Alignment) type {
         /// Asserts that the new length is less than or equal to the previous length.
         pub fn shrinkRetainingCapacity(self: *Self, new_len: usize) void {
             assert(new_len <= self.items.len);
+            @memset(self.items[new_len..], undefined);
             self.items.len = new_len;
         }
 
+        /// Reduce length to 0.
         /// Invalidates all element pointers.
         pub fn clearRetainingCapacity(self: *Self) void {
+            @memset(self.items, undefined);
             self.items.len = 0;
         }
 
-- 
2.52.0

